mypy/semanal.py

```diff
@@ -719,9 +719,48 @@ def analyze_class_body(self, defn: ClassDef) -> Iterator[bool]:
                 yield True
                 self.calculate_abstract_status(defn.info)
                 self.setup_type_promotion(defn)
-
+                self.apply_class_plugin_hooks(defn)
                 self.leave_class()
 
+    def apply_class_plugin_hooks(self, defn: ClassDef) -> None:
+        """Apply a plugin hook that may infer a more precise definition for a class."""
+
+        def get_fullname(expr: Expression) -> Optional[str]:
+            if isinstance(expr, CallExpr):
+                return get_fullname(expr.callee)
+            elif isinstance(expr, (MemberExpr, NameExpr)):
+                if expr.fullname:
+                    return expr.fullname
+
+                # If we don't have a fullname look it up.
+                sym = self.lookup_type_node(expr)
```

Comment:
Add a longer comment here explaining that currently base classes are analysed in different manner than other expressions (analysed by ``exprtotype.py`` etc.) and therefore some AST nodes doesn't have names set. Maybe also add a ``TODO`` item to unify this (this is however low priority since this will require major rewriting with minor benefits).